models/objects.py

```diff
@@ -0,0 +1,284 @@
+
+            alpha = ''
+            for i in range(need_letters):
+                alpha += chr(((replacements['number']) // 26 ** i) % 26 + ord('a'))
+
+            alpha = alpha[::-1]  # reverse the string
+            replacements['alpha'] = alpha
+
+            # generate the ALPHA part
+            replacements['ALPHA'] = alpha.upper()
+
+            for key in ['instrument', 'yyyy', 'yy', 'mm', 'dd', 'alpha', 'ALPHA', 'number']:
+                fmt = fmt.replace(f'<{key}>', str(replacements[key]))
+
+            return fmt
+
+        return partial(name_func, fmt=format_string)
+
+    @staticmethod
+    def get_last_id_for_naming(convention, present_time=None, session=None):
+        """Get the ID of the last object before the given date (defaults to now).
+
+        Will query the database for an object with a created_at which is the last before
+        the start of this year, month or day (depending on what exists in the naming convention).
+        Will return the ID of that object, or 0 if no object exists.
+
+        Parameters
+        ----------
+        convention: str
+            The naming convention that will be used to generate the name.
+            Example: SomeText_<instrument>_<yyyy><mm>_<number>
+        present_time: datetime.datetime, optional
+            The time to use as the present time. Defaults to now.
+        session: sqlalchemy.orm.session.Session, optional
+            The session to use for the query. If not given, will open a new session
+            that will be automatically closed at the end of this function.
+
+        Returns
+        -------
+        int
+            The ID of the last object before the given date.
+        """
+        if present_time is None:
+            present_time = datetime.datetime.utcnow()
+
+        # figure out what the time frame should be:
+        if '<yyyy>' in convention:
+            start_time = datetime.datetime(present_time.year, 1, 1)
+        elif '<mm>' in convention:
+            start_time = datetime.datetime(present_time.year, present_time.month, 1)
+        elif '<dd>' in convention:
+            start_time = datetime.datetime(present_time.year, present_time.month, present_time.day)
+        else:
+            return 0  # if none of these format specifiers are present, just assume there is no last object
+
+        with SmartSession(session) as session:
+            last_obj = session.scalars(
+                sa.select(Object).where(Object.created_at < start_time).order_by(Object.created_at.desc())
+            ).first()
+            if last_obj is None:
+                return 0
+            return last_obj.id
+
+
+# add an event listener to catch objects before insert and generate a name for them
+@sa.event.listens_for(Object, 'before_insert')
+def generate_object_name(mapper, connection, target):
+    if target.name is None:
+        target.name = 'placeholder'
+
+
+@sa.event.listens_for(sa.orm.session.Session, 'after_flush_postexec')
```

Comment:
The only way I got this to work is using an event listener that does trigger additional updates (according to the docs). 
The way it works is: 
a) the object gets flushed to DB, at which point it gets an ID from the DB autoincrement. 
b) the event triggers and seeing it has a "placeholder" name it would use the ID to create a new name. 
c) the new name triggers a further update that needs to be flushed (it will be committed or not)
d) this triggers the event listener again, but it does nothing since the name is no longer "placeholder". 

Yes, I've spent way too long on this.

Reply:
Also, notice that I'm testing this explicitly in `test_objects`

Reply:
OK, sounds good.